device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * This is the base class for DeviceClientManager and MultiplexingClientManager
+ * The shared logic for reconnection and handling status change callback exists here.
+ */
+@Slf4j
+public abstract class ClientManagerBase implements IotHubConnectionStatusChangeCallback, ConnectionStatusTracker
+{
+    protected static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 2;
+
+    // Initialize the connection status as DISCONNECTED
+    protected ConnectionStatus lastKnownConnectionStatus = ConnectionStatus.DISCONNECTED;
+
+    // Tracks connection status of the dependency connection (multiplexing client connection status for example in this case for DeviceClientManager)
+    protected ConnectionStatusTracker dependencyConnectionStatusTracker;
+
+    public abstract void openClient() throws IOException;
+    public abstract void closeClient() throws IOException;
+    public abstract String getClientId();
+
+    // Since the client manager is in charge of handling the connection status callback, this method is a no-op
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
+    {
+        return;
+    }
+
+    public void closeNow()
+    {
+        synchronized (lastKnownConnectionStatus)
+        {
+            try
+            {
+                log.debug("Closing the client instance " + getClientId() + " ...");
+                closeClient();
+            }
+            catch (IOException ex)
+            {
+                log.error("Exception thrown while closing client instance " + getClientId() + " : ", ex);
+            }
+            finally
+            {
+                // Once the connection is closed, set the Status to DISCONNECTED
+                lastKnownConnectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+    }
+
+    // When client manager is being opened it first makes sure the client is in a DISCONNECTED state
+    // If the client is in CONNECTING or CONNECTED state, Open will be no-op.
+    public void open() throws IOException
+    {
+        synchronized (lastKnownConnectionStatus)
+        {
+            // Do not attempt to CONNECT if the connection status is not DISCONNECTED. This ensure that only one process is going to attempt to connect
+            if (lastKnownConnectionStatus != ConnectionStatus.DISCONNECTED)
+            {
+                return;
+            }
+            else
+            {
+                // Set the connection status to CONNECTING
+                lastKnownConnectionStatus = ConnectionStatus.CONNECTING;
+            }
+        }
+
+        establishConnection();
+    }
+
+    @Override
+    public ConnectionStatus getConnectionStatus(){
+        return this.lastKnownConnectionStatus;
+    }
+
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext)
```

Comment:
Let's avoid the verb execute. It doesn't pass PoliCheck.

Reply:
I don't think we can avoid it, it's in our SDK interface and we have to implement it :(

Reply:
Which SDK? The legacy DeviceClient or the new MultiplexingClient?